armory/eval/evaluator.py

```diff
@@ -92,10 +92,24 @@ def _gather_env_variables(self):
         if not self.armory_global_config["verify_ssl"]:
             self.extra_env_vars["VERIFY_SSL"] = "false"
 
+        cuda_var_exists = (
+            True if os.getenv("CUDA_VISIBLE_DEVICES") is not None else False
+        )
+        # Existing value should override armory flags
+
+        # Set visible gpus
         if self.config["sysconfig"].get("use_gpu", None):
             gpus = self.config["sysconfig"].get("gpus")
             if gpus is not None:
-                self.extra_env_vars["NVIDIA_VISIBLE_DEVICES"] = gpus
+                if self.no_docker and not cuda_var_exists:
+                    self.extra_env_vars["CUDA_VISIBLE_DEVICES"] = gpus
+                if not self.no_docker:
+                    self.extra_env_vars["NVIDIA_VISIBLE_DEVICES"] = gpus
```

Comment:
Would be useful to at a minimum show a warning (or throw error) when CUDA_VISIBLE_DEVICES is already set but a gpu flag is passed in as running `CUDA_VISIBLE_DEVICES=3 python3 -m armory run scenario_configs/cifar10_baseline.json --no-docker --gpus 1` will load everything onto gpu 3 without any warnings.

Change below tested to work locally.


```suggestion
                if self.no_docker and not cuda_var_exists:
                    self.extra_env_vars["CUDA_VISIBLE_DEVICES"] = gpus
                elif not self.no_docker:
                    self.extra_env_vars["NVIDIA_VISIBLE_DEVICES"] = gpus
                else:
                    log.warning(
                        "CUDA_VISIBLE_DEVICES already set, ignoring gpus argument"
                    )
```

Reply:
I've pulled the warning up a level so it will be logged whether use_gpu is true or false.